### ISSUE TITLE
Panels in Windowbuilder Design tab are empty

### ISSUE THREAD

**Issue by drchilton** [eclipse-windowbuilder/windowbuilder#747]
One GUI class in my application displays empty panels in the Design tab.  The Design tab opens and does not display any errors.  It looks like this:
![WindowBuilderDesignTab](https://github.com/eclipse-windowbuilder/windowbuilder/assets/5721445/09523872-4dfa-4468-a429-4c57da08364e)
The code is in the attached file.  I left out the import statements, but if needed I can add them.
[GearRatioView.txt](https://github.com/eclipse-windowbuilder/windowbuilder/files/14949972/GearRatioView.txt)

This class obtains data from other classes that communicate with a database and uses a renderer to format the output in a JTable. Additional other classes are used as entity classes to hold the data during processing.   I did not include these but can upload them if needed.  The output (the application runs as it should) displays the data in a formatted table as shown below.  While some of the formatting may look peculiar, I can assure you that it is running properly.
![image](https://github.com/eclipse-windowbuilder/windowbuilder/assets/5721445/35d38695-f084-4df2-a2e8-6ca7a761173a)
I have another class in the same project that does not exhibit this behavior but displays in the Design tab correctly.  The difference in the two classes is mostly the table in this one--the other one does not have a table.
I am using eclipse 2024-03, Windowbuilder 1.15.0 and I've tried this with JavaSE-1.8 and JavaSE-17.

**Comment by ptziegler** [comment]
Even without the additional classes, I am able to open the widget without error. So I don't believe there is anything wrong with the example you've provided.

![image](https://github.com/eclipse-windowbuilder/windowbuilder/assets/70652978/1b3426d5-96d6-40f1-b4b8-97c3738861f6)

Given that not even the palette is rendered, I assume that an exception is thrown internally. If you the error view by clicking "Window" -> "Show View" -> "Error Log" in the main menu, is there anything logged?

**Comment by drchilton** [comment]
Thanks for responding so quickly.  I just opened the Design view and the error log shows two (identical) entries for that time:

Unhandled event loop exception             org.eclipse.ui                     4/12/24, 4:36 PM
Unhandled event loop exception             org.eclipse.ui                     4/12/24, 4:36 PM

I looked this up on Stack Overflow and they were talking about making changes to the eclipse.ini file.  I did as was suggested and then eclipse wouldn’t run at all, so I removed the change and now it starts again, but with the same result—blank panels in the Design tab.

Here is my eclipse.ini file in case that might help:

-startup
plugins/org.eclipse.equinox.launcher_1.6.700.v20240213-1244.jar
--launcher.library
C:\Users\mchil\.p2\pool\plugins\org.eclipse.equinox.launcher.win32.win32.x86_64_1.2.900.v20240213-1244
-product
org.eclipse.epp.package.java.product
-showsplash
C:\Users\mchil\.p2\pool\plugins\org.eclipse.epp.package.common_4.31.0.20240307-1200
--launcher.defaultAction
openFile
--launcher.appendVmargs
-vmargs
-Dorg.eclipse.ecf.provider.filetransfer.excludeContributors=org.eclipse.ecf.provider.filetransfer.httpclientjava
-Dosgi.requiredJavaVersion=17
***@***.******@***.***/eclipse-workspace>
-Dosgi.dataAreaRequiresExplicitInit=true
-Dorg.eclipse.swt.graphics.Resource.reportNonDisposed=true
-Declipse.e4.inject.javax.warning=false
-Dsun.java.command=Eclipse
-Xms512m
-Xmx2048m
-XX:+UseG1GC
-XX:+UseStringDeduplication
--add-modules=ALL-SYSTEM
-Djava.security.manager=allow
-Declipse.p2.max.threads=10
-Doomph.update.url=https://download.eclipse.org/oomph/updates/milestone/latest
-Doomph.redirection.index.redirection=index:/->https://raw.githubusercontent.com/eclipse-oomph/oomph/master/setups/

Thanks for assisting.

Michael A. Chilton, PhD
Associate Professor Emeritus
Kansas State University
Manhattan, KS 66506
***@***.***

**Comment by ptziegler** [comment]
> Thanks for responding so quickly.  I just opened the Design view and the error log shows two (identical) entries for that time:

> Unhandled event loop exception             org.eclipse.ui                     4/12/24, 4:36 PM
> Unhandled event loop exception             org.eclipse.ui                     4/12/24, 4:36 PM

If you right-click on the entry, you should be able to "copy" the error, which should contain a full stack trace. Can you post this complete error message?

**Comment by drchilton** [comment]
I’ve attached the stack traces for the two errors I’m seeing.

One thing I should mention is that when I comment out the table model, WindowBuilder works just as you saw it, but of course, with no data.

Michael A. Chilton, PhD
Associate Professor Emeritus
Kansas State University
Manhattan, KS 66506
***@***.***

**Comment by ptziegler** [comment]
Thanks! That already tracks down the problem by quite a lot!

> One thing I should mention is that when I comment out the table model, WindowBuilder works just as you saw it, but of course, with no data.

Hm... that matches the stack trace you've provided. For some reason, the array returned by `calculatedGearRatios.calcGearRatios(internalRatios, frontSprockets, rearSprockets)` is `null`, as opposed to containing the table entries.

Would it be possible to provided the remaining Java classes? At this point I'm certain that something goes wrong when processing the Java code. So being able to see what's actually happening would help quite a lot.

**Comment by drchilton** [comment]
[BicycleGears.zip](https://github.com/eclipse-windowbuilder/windowbuilder/files/14962275/BicycleGears.zip)
I’ve attached a zip file with the entire project.  There are 4 packages: controller, model, test and ui.  You can ignore the classes in the test package.  There are two files, DbConfig.java and DbQueries.java, that contain the database connection information and the queries, but I don’t think you’ll need any of the database data.  At this stage of development, the data is hardcoded into the entity classes.  There is one exception and that involves duplicate gear ratios.  Rather than include the data here, I think it would be easier for you to comment out line 270 [ “formatTable();” ] in the GearRatioView.java class, since the behavior of WindowBuilder is the same whether you format the data or not.

Two other files, HashMapDuplicates and TableCellBackgroundRenderer in the controller package are also not used.

The way the program works is that it uses the three distinct sets of data, internal gear ratios, front wheel tooth count and rear wheel tooth count to calculate the ratios..  There are 3 internal gears, 3 front wheel sprockets and 9 rear wheel sprockets.  This yields 81 total gear ratios.  These are displayed in the table.  Many of the ratios repeat themselves so these repetitions are colored the same, but each repetition has its own color.  The first and last columns in the table display the rear tooth count.  The headers are “”Rear teeth” for those columns and the others are the front teeth counts.  These repeat three times—once for each internal gear.

**Comment by ptziegler** [comment]
Awesome! I can reproduce the error on my side and was able to isolate the problem to the following line:

> ratios[i][j] = frontToothCount[j] / rearToothCount[k] * internalRatios[i];

The `rearToothCount` array is -for some reason- not initialized properly. So you end up with `frontToothCount[j] / 0 * internalRatios[i];`, leading to a division by zero. I have to further investigate why this is the case, though.
`

**Comment by ptziegler** [comment]
Correction: It's a division by `null`, not `0`.

> java.lang.NullPointerException: Cannot invoke "java.lang.Double.doubleValue()" because "internalRatios[i]" is null

**Comment by ptziegler** [comment]
The following lines are not processed correctly:

```
internalRatios[0] = lowRatio; 
internalRatios[1] = middleRatio;
internalRatios[2] = highRatio;
```

As a workaround, you can replace:
> private Double[] internalRatios = new Double[3];

with 
> private Double[] internalRatios;

and

> internalRatios[0] = lowRatio; 
> internalRatios[1] = middleRatio;
> internalRatios[2] = highRatio;

with

> internalRatios = new Double[] { lowRatio, middleRatio, highRatio };

[GearRatioView.java.txt](https://github.com/eclipse-windowbuilder/windowbuilder/files/14964487/GearRatioView.java.txt)

With this change, the widget should be shown in the design tab and the table should contain the correct values:

![image](https://github.com/eclipse-windowbuilder/windowbuilder/assets/70652978/ab891a65-2ee8-4a45-adc7-350fe496716e)

**Comment by drchilton** [comment]
Right as rain!  Nice job.  It works now.  My only question is why did the application run without problems, but it tripped up in WindowBuilder?  Anyway, thanks for your help! I really appreciate it.

**Comment by ptziegler** [comment]
Fixed via #766.

The problem is that WindowBuilder ignores any modifications to an array that are done outside of the initial declaration.

So instead of:

```
Double[] internalRatios = new Double[3];
internalRatios[0] = lowRatio; 
internalRatios[1] = middleRatio;
internalRatios[2] = highRatio;
spinnerLowInternal.setValue(internalRatios[0]);
```

WindowBuilder only "sees":

```
Double[] internalRatios = new Double[3];
spinnerLowInternal.setValue(internalRatios[0]);
```

Meaning the array remains uninitialized. This is also the reason why this problem was fixed after the assignments were moved (`i.e. internalRatios = new Double[] {...}`). I've tested it with the project you provided and now the table is rendered, even without any code adjustments.

![image](https://github.com/eclipse-windowbuilder/windowbuilder/assets/70652978/8a1f597f-c17c-4cc3-ad09-48801ed8482e)
